packages/flutter_tools/lib/src/base/process_manager.dart

```diff
@@ -479,3 +492,296 @@ class _RecordingProcess implements Process {
+///   invocation:
+///   - `pid` (required): The process id integer.
+///   - `basename` (required): A string specifying the base filename from which
+///     the incovation's `stdout` and `stderr` files can be located.
+///   - `executable` (required): A string specifying the path to the executable
+///     command that kicked off the process.
+///   - `arguments` (required): A list of strings that were passed as arguments
+///     to the executable.
+///   - `workingDirectory` (required): The current working directory from which
+///     the process was spawned.
+///   - `environment` (required): A map from string environment variable keys
+///     to their corresponding string values.
+///   - `mode` (optional): A string specifying the [ProcessStartMode].
+///   - `stdoutEncoding` (optional): The name of the encoding scheme that was
+///     used in the `stdout` file. If unspecified, then the file was written
+///     as binary data.
+///   - `stderrEncoding` (optional): The name of the encoding scheme that was
+///     used in the `stderr` file. If unspecified, then the file was written
+///     as binary data.
+///   - `exitCode` (required): The exit code of the process, or null if the
+///     process was not responding.
+///   - `daemon` (optional): A boolean indicating that the process is to stay
+///     resident during the entire lifetime of the master Flutter tools process.
+/// - a `stdout` file for each process invocation. The location of this file
+///   can be derived from the `basename` manifest property like so:
+///   `'$basename.stdout'`.
+/// - a `stderr` file for each process invocation. The location of this file
+///   can be derived from the `basename` manifest property like so:
+///   `'$basename.stderr'`.
+class ReplayProcessManager implements ProcessManager {
+  final List<Map<String, dynamic>> _manifest;
+  final Directory _dir;
+
+  ReplayProcessManager._(this._manifest, this._dir);
+
+  /// Creates a new `ReplayProcessManager` capable of replaying a recording at
+  /// the specified location.
+  ///
+  /// If [location] represents a file, it will be treated like a recording
+  /// ZIP file. If it points to a directory, it will be treated like an
+  /// unzipped recording. If [location] points to a non-existent file or
+  /// directory, an [ArgumentError] will be thrown.
+  static Future<ReplayProcessManager> create(String location) async {
+    Directory dir;
+    switch (await FileSystemEntity.type(location)) {
+      case FileSystemEntityType.FILE:
+        dir = await Directory.systemTemp.createTemp('flutter_tools_');
+        os.unzip(new File(location), dir);
+        addShutdownHook(() async {
+          await dir.delete(recursive: true);
+        });
+        break;
+      case FileSystemEntityType.DIRECTORY:
+        dir = new Directory(location);
+        break;
+      case FileSystemEntityType.NOT_FOUND:
+        throw new ArgumentError.value(location, 'location');
```

Comment:
ArgumentError takes a message parameter, you could use it to help make this failure case clearer.